File: src/main/java/ee/tenman/portfolio/alphavantage/AlphaVantageResponse.java
package ee.tenman.portfolio.alphavantage;

import com.fasterxml.jackson.annotation.JsonProperty;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.math.BigDecimal;
import java.util.TreeMap;

@Data
@Builder
@AllArgsConstructor
@NoArgsConstructor
public class AlphaVantageResponse {
  @JsonProperty("Monthly Time Series")
  private TreeMap<String, AlphaVantageDayData> monthlyTimeSeries;

  @JsonProperty("Information")
  private String information;

  @JsonProperty("Error Message")
  private String errorMessage;

  @Data
  @Builder
  @AllArgsConstructor
  @NoArgsConstructor
  public static class AlphaVantageDayData {
    @JsonProperty("4. close")
    private BigDecimal close;
  }

}

File: src/main/kotlin/ee/tenman/portfolio/alphavantage/AlphaVantageResponse.kt
package ee.tenman.portfolio.alphavantage

import com.fasterxml.jackson.annotation.JsonProperty
import ee.tenman.portfolio.common.DailyPriceData
import java.math.BigDecimal
import java.util.TreeMap

data class AlphaVantageResponse(
  @JsonProperty("Monthly Time Series")
  val monthlyTimeSeries: TreeMap<String, AlphaVantageDailyPriceData>? = null,
  @JsonProperty("Time Series (Daily)")
  val dailyTimeSeries: TreeMap<String, AlphaVantageDailyPriceData>? = null,
  @JsonProperty("Information")
  val information: String? = null,
  @JsonProperty("Error Message")
  val errorMessage: String? = null,
) {
  data class AlphaVantageDailyPriceData(
    @JsonProperty("1. open")
    override val open: BigDecimal,
    @JsonProperty("2. high")
    override val high: BigDecimal,
    @JsonProperty("3. low")
    override val low: BigDecimal,
    @JsonProperty("4. close")
    override val close: BigDecimal,
    @JsonProperty("5. volume")
    override val volume: Long,
  ) : DailyPriceData
}
